container/new_push.bzl

```diff
@@ -0,0 +1,166 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An new implementation of container_push based on google/containerregistry using google/go-containerregistry.
+
+This wraps the rules_docker.container.go.cmd.pusher.pusher executable in a
+Bazel rule for publishing images.
+"""
+
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("@io_bazel_rules_docker//container:providers.bzl", "PushInfo")
+load(
+    "//container:layer_tools.bzl",
+    _layer_tools = "tools",
+)
+load(
+    "//skylib:path.bzl",
+    "runfile",
+)
+
+def _get_runfile_path(ctx, f):
+    return "${RUNFILES}/%s" % runfile(ctx, f)
+
+def _impl(ctx):
+    """Core implementation of new_container_push."""
+
+    # TODO (xiaohegong): 1) Possible optimization for efficiently pushing intermediate
+    # representation, similar with the old python implementation, e.g., push-by-layer.
+    # Some of the digester arguments omitted from before: --tarball, --config, --manifest, --digest, --layer, --oci.
+    # 2) The old implementation outputs a {image_name}.digest for compatibility with container_digest, omitted for now.
+    # 3) Use and implementation of attr.stamp.
+
+    pusher_args = []
+
+    # Parse and get destination registry to be pushed to
+    registry = ctx.expand_make_variables("registry", ctx.attr.registry, {})
+    repository = ctx.expand_make_variables("repository", ctx.attr.repository, {})
+    tag = ctx.expand_make_variables("tag", ctx.attr.tag, {})
+
+    # If a tag file is provided, override <tag> with tag value
+    runfiles_tag_file = []
+    if ctx.file.tag_file:
+        tag = "$(cat {})".format(_get_runfile_path(ctx, ctx.file.tag_file))
+        runfiles_tag_file = [ctx.file.tag_file]
+
+    pusher_args += ["-dst", "{registry}/{repository}:{tag}".format(
+        registry = registry,
+        repository = repository,
+        tag = tag,
+    )]
+
+    # Find and set src to correct paths depending the image format to be pushed
+    if ctx.attr.format == "oci":
+        found = false
+        for f in ctx.files.image:
+            if f.basename == "index.json":
+                pusher_args += ["-src", "{index_dir}".format(
+                    index_dir = f.dirname,
+                )]
+                found = True
+        if not found:
+            fail("Did not find an index.json in the image attribute specified to {}".format(ctx.attr.image))
+    elif len(ctx.files.image) > 0:
```

Comment:
We should check the number of images provided is exactly one when format is `docker`.
Instead of having the logic for format `docker` under the elif & else clauses, I would move this under a separate if condition for better readability. So something like:
```
if ctx.attr.format == "docker":
   if len(ctx.files.image) == 0:
       fail("Attribute image {} to {} did not contain an image tarball".format(ctx.attr.image, ctx.label))
   if len(ctx.files.images) > 1:
       fail("Attribute image {} to {} had {} files. Expected exactly 1".format(ctx.attr.image, ctx.label, len(ctx.files.images))
   pusher_args += ["-src", str(ctx.files.image[0].path)]
```